README.md
[What is Pytim](#what-is-pytim) | [Examples](#example) | [More info](#more-info)  | [How to Install](#installation) | [References](#references)


[![Build Status](https://travis-ci.org/Marcello-Sega/pytim.svg?branch=master)](https://travis-ci.org/Marcello-Sega/pytim)
[![GitHub tags](https://img.shields.io/github/tag/Marcello-Sega/pytim.svg)](https://github.com/Marcello-Sega/pytim/)
[![GitHub issues](https://img.shields.io/github/issues/Marcello-Sega/pytim.svg)](https://github.com/Marcello-Sega/pytim/issues)
[![codecov](https://codecov.io/gh/Marcello-Sega/pytim/branch/master/graph/badge.svg)](https://codecov.io/gh/Marcello-Sega/pytim)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/5f3f2e7be75b46c1a6e4a3d44e3bb900)](https://www.codacy.com/app/Marcello-Sega/pytim?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=Marcello-Sega/pytim&amp;utm_campaign=Badge_Grade)
[![Code Climate](https://codeclimate.com/github/Marcello-Sega/pytim/badges/gpa.svg)](https://codeclimate.com/github/Marcello-Sega/pytim)
[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)

<sub>**Disclaimer**: Pytim is in **beta-stage** right now and while a systematic testing system has been set up, this has not yet total coverage. We will keep rolling out more examples and, still, some new features. If you try this software out and have some suggestions, remarks, or bugfixes, feel free to comment here on GitHub and/or make a pull request. </sub>

 <sub>**News**: The paper about pytim has been published on J. Comput. Chem. It is open access and you can [download the pdf](https://onlinelibrary.wiley.com/doi/epdf/10.1002/jcc.25384) from Wiley <img  src="https://licensebuttons.net/l/by-nc/4.0/80x15.png"> (see also the [references](#refs))

# What is Pytim

[Pytim](https://marcello-sega.github.io/pytim/) is a cross-platform python implementation of several methods for the detection of fluid interfaces in molecular simulations. It is based on [`MDAnalysis`](https://www.mdanalysis.org/), but it integrates also seamlessly with `MDTraj`, and can be even used for *online* analysis during an `OpenMM` simulation (see further down for examples [with `MDTraj`](#mdtraj-example) and [with `OpenMM`](#openmm-example)).

So far the following interface/phase identification methods have been implemented:
<img src="https://github.com/Marcello-Sega/pytim/raw/IMAGES/_images/micelle_cut.png" width="380" align="right" style="z-index:999;">


* ITIM
* GITIM 
* SASA
* Willard Chandler
* Chacon Tarazona
* DBSCAN filtering

## Supported formats
Pytim relies on the [MDAnalysis](http://www.mdanalysis.org/)
package for reading/writing trajectories, and work therefore seamlessly for a number of popular trajectory formats, including:  
* GROMACS
* CHARMM/NAMD
* LAMMPS
* AMBER
* DL_Poly

as well as common structure file formats such as XYZ or PDB (have a look at the [complete list](https://pythonhosted.org/MDAnalysis/documentation_pages/coordinates/init.html#id1))

## Install from PyPi or Anaconda - [![PyPI version](https://badge.fury.io/py/pytim.svg)](https://badge.fury.io/py/pytim) [![Anaconda-Server Badge](https://anaconda.org/conda-forge/pytim/badges/version.svg)](https://anaconda.org/conda-forge/pytim)

PyPi:     ``` pip install --user --upgrade pytim ```

Anaconda: ``` conda install -c conda-forge pytim ```



# <a name="example"></a> Show me an example usage, now!

Ok, ok ... have a look below: this example is about computing molecular layers of a flat interface:

### Step 1: interface identification


```python
import MDAnalysis as mda
import pytim
from pytim.datafiles import WATER_GRO

# load the configuration in MDAnalysis
u = mda.Universe(WATER_GRO)

# compute the interface using ITIM. Identify 4 layers.
inter = pytim.ITIM(u,max_layers=4)
```

### That's it. There's no step 2!

Now interfacial atoms are accessible in different ways, pick the one you like:

1. Through the `atoms` group accessible as

```python
inter.atoms.positions # this is a numpy array holding the position of atoms in the layers
```

    array([[ 22.10000038,  16.05999947,  94.19633484],
           [ 22.43999863,  16.97999954,  93.96632385],
           ...,
           [ 33.70999908,  49.02999878,  62.52632904],
           [ 34.06999969,  48.18000031,  61.16632843]], dtype=float32)

2. Using the label that each atom in the `MDAnalysis` universe now has, which specifies in which layer it is found: 

```python
u.atoms.layers  # -1 if not in any layer, 1 if in the first layer, ...
```

3. Using the layers groups, stored as a list (of lists, in case of upper/lower layers in flat interfaces) of groups: 

```pythin
inter.layers

array([[<AtomGroup with 780 atoms>, <AtomGroup with 690 atoms>,
        <AtomGroup with 693 atoms>, <AtomGroup with 660 atoms>],
       [<AtomGroup with 777 atoms>, <AtomGroup with 687 atoms>,
        <AtomGroup with 663 atoms>, <AtomGroup with 654 atoms>]], dtype=object)
```

## Visualisation

Pytim can export in different file formats: the `PDB` format with the `beta` field used to tag the layers, `VTK`, `cube` for both continuous surfaces and particles, and, of course, all formats supported by `MDAnalysis`. 

In [`VMD`](www.ks.uiuc.edu/Research/vmd/), for example, using `beta == 1` allows you to select all atoms in the first interfacial layer. Just save your `PDB` file with layers information using

```python
inter.writepdb('myfile.pdb')
```


In a `jupyter` notebook, you can use `nglview` like this:


```python
import nglview
v = nglview.show_mdanalysis(u)
v.camera='orthographic'
v.center()
system = v.component_0
colors = ['','red','orange','yellow','white']

for n in [1,2,3,4]:
    system.add_spacefill(selection = inter.atoms[inter.atoms.layers==n].indices, color=colors[n] )

system.add_spacefill(selection = (u.atoms - inter.atoms).indices, color='gray' )
```
```python
# this must go in a separate cell
v.display()
```
<p align="center">

<img src="https://github.com/Marcello-Sega/pytim/raw/IMAGES/_images/output_13_0.png" width="60%" align="center" style="z-index:999;">
</p>

## Analysing trajectories (`MDAnalysis` and `mdtraj`)

Once one of the pytim classes (`ITIM`,`GITIM`,`WillardChandler`,...) has been initialised, whenever a new frame is loaded, the interfacial properties will be calculated automatically without need of doing anything else

### MDAnalysis example

```python
import MDAnalysis as mda
import pytim 
from pytim.datafiles import WATER_GRO, WATER_XTC 

u = mda.Universe(WATER_GRO,WATER_XTC)
inter = pytim.ITIM(u)
for step in u.trajectory[:]:
    print "surface atoms:", repr(inter.atoms)
```

### mdtraj example

Under the hood, `pytim` uses `MDAnalysis`, but this is made (almost completely) transparent to the user, so that interoperability with other software is easy to implement. For example, to analyse a trajectory loaded with [`MDTraj`](https://mdtraj.org), it is enough to do the following:

```python
import mdtraj
import pytim                     
from pytim.datafiles import WATER_GRO, WATER_XTC 

t = mdtraj.load_xtc(WATER_XTC,top=WATER_GRO) 
inter = pytim.ITIM(t) 
for step in t[:]:
        print "surface atoms:" , repr(inter.atoms.indices)
```

### openmm example

Another example is using `pytim` to perform *online* interfacial analysis during an [`OpenMM`](https://openmm.org/) simulation:
```python
# openmm imports
from simtk.openmm.app import *
from simtk.openmm import *
from simtk.unit import *
# pytim
import pytim
from pytim.datafiles import WATER_PDB

# usual openmm setup, we load one of pytim's example files
pdb = PDBFile(WATER_PDB)
forcefield = ForceField('amber99sb.xml', 'spce.xml')
system = forcefield.createSystem(pdb.topology, nonbondedMethod=PME,
        nonbondedCutoff=1*nanometer)
integrator = LangevinIntegrator(300*kelvin, 1/picosecond, 0.002*picoseconds)
simulation = Simulation(pdb.topology, system, integrator)
simulation.context.setPositions(pdb.positions)

# just pass the openmm Simulation object to pytim
inter = pytim.ITIM(simulation)
print repr(inter.atoms)

# the new interfacial atoms will be computed at the end
# of the integration cycle
simulation.step(10)
print repr(inter.atoms)

```



## <a name="non-flat-interfaces"></a> What if the interface is not flat? 

You could use GITIM, or the Willard-Chandler method. 
Let's have a look first at **GITIM**:

```python
import MDAnalysis as mda
import pytim
from   pytim.datafiles import *

u       = mda.Universe(MICELLE_PDB)
g       = u.select_atoms('resname DPC')

interface = pytim.GITIM(u,group=g,molecular=False,
                        symmetry='spherical',alpha=2.5)
layer = interface.layers[0]
interface.writepdb('gitim.pdb',centered=False)
print repr(layer)
<AtomGroup with 872 atoms>
```

`nglview` can be used to show a section cut of the micelle in a `jupyter` notebook:

```python
import nglview
import numpy as np

v = nglview.show_mdanalysis(u)
v.camera='orthographic'
v.center()
v.add_unitcell()
system = v.component_0
system.clear()

selection = g.atoms.positions[:,2]>30.
system.add_spacefill(selection =g.atoms.indices[selection])

selection = np.logical_and(inter.atoms.layers==1,inter.atoms.positions[:,2]>30.)
system.add_spacefill(selection =inter.atoms.indices[selection], color='yellow' )
```

```python
v.display()
```

<p align="center">

<img src="https://github.com/Marcello-Sega/pytim/raw/IMAGES/_images/micelle-gitim.png" width="40%" align="middle">
</p>

The **Willard-Chandler** method can be used, instead to find out isodensity surfaces:

```python
import MDAnalysis as mda
import pytim
from pytim.datafiles import MICELLE_PDB
import nglview

u = mda.Universe(MICELLE_PDB)
g = u.select_atoms('resname DPC')
```

```python
interface = pytim.WillardChandler(u, group=g, mesh=1.5, alpha=3.0)
interface.writecube('data.cube')
```

Done, the density file is written in `.cube` format, we can now open it with 
programs such as [`Paraview`](https://www.paraview.org/), or visualize it in a
jupyter notebook with `nglview`

```python
view = nglview.show_mdanalysis(u.atoms) # the atoms, this will be component_0 in nglview
view.add_component('data.cube') # the density data, this will be component_1 in nglview

view.clear() # looks like this is needed in order for view._display_image() to work correctly 
# let's center the view on our atoms, and draw them as spheres  
view.component_0.center()
view.component_0.add_spacefill(selection='DPC')

# let's add a transparent, red representation for the isodensity surface at 50% density
view.component_1.add_surface(color='red',isolevelType="value",isolevel=0.5,opacity=0.6) 

# add a nice simulation box
view.add_unitcell()
```

```python
view.display()
```

<p align="center">

<img src="https://github.com/Marcello-Sega/pytim/raw/IMAGES/_images/micelle-willard-chandler.png" width="60%" align="middle">
</p>

### Calculate multiple layers with GITIM: solvation shells of glucose
```python
import MDAnalysis as mda
import pytim
from   pytim.datafiles import GLUCOSE_PDB

u       = mda.Universe(GLUCOSE_PDB)
solvent = u.select_atoms('name OW')
glc     = u.atoms - solvent.residues.atoms

inter = pytim.GITIM(u, group=solvent, molecular=True,
                    max_layers=3, alpha=2)

for i in [0,1,2]:
    print "Layer "+str(i),repr(inter.layers[i])
```
```python
Layer 0 <AtomGroup with 81 atoms>
Layer 1 <AtomGroup with 186 atoms>
Layer 2 <AtomGroup with 240 atoms>
```

```python
import nglview
import numpy as np

v = nglview.show_mdanalysis(u)
v.camera='orthographic'
v.center()
v.add_unitcell()

v.clear()

# glucose
v.add_licorice(selection =glc.atoms.indices,radius=.35)

colors = ['yellow','blue','purple']
# hydration layers
for layer in [0,1,2]:
    selection = np.logical_and(inter.atoms.layers==layer+1 ,inter.atoms.positions[:,2]>9)
    v.add_spacefill(selection =inter.atoms.indices[selection], color=colors[layer] )

```
```python
v.display()
```
<p align="center">

<img src="https://github.com/Marcello-Sega/pytim/raw/IMAGES/_images/glc-gitim.png" width="60%" align="middle">
</p>

When calculating surfaces with `GITIM`, it can happen that several disconnected, closed surfaces are found in a simulation box. To restrict the analysis to the largest, clustered interfacial atoms (also when calculating multiple layers), one can pass the `biggest_cluster_only` option, as in:

```python
inter = pytim.GITIM(u, group=solvent, molecular=True, max_layers=3, alpha=2, 
                    biggest_cluster_only=True, cluster_cut = 3.5)
```
In order for this option to have any effect, a `cluster_cut` value should also be passed.

# More info

Have a look at some jupyter notebooks:

1. [An introduction to Pytim](https://github.com/Marcello-Sega/pytim/blob/master/notebooks/An%20introduction%20to%20Pytim.ipynb) 
2. [The Willard-Chandler method]( https://github.com/Marcello-Sega/pytim/blob/master/notebooks/Willard-Chandler%20and%20Cube%20format.ipynb)

Browse the examples in the online manual:

3. [Pytim Online Manual](https://marcello-sega.github.io/pytim/quick.html)

Check out the Pytim Poster from the 10th Liquid Matter Conference 

4. [Available on ResearchGate](http://dx.doi.org/10.13140/RG.2.2.18613.17126)  DOI:10.13140/RG.2.2.18613.17126


# <a name="installation"></a> How to install the package and the documentation? 

## From the PyPI

this will install the latest release present in the Python Package Index:

```
pip install --user --upgrade pytim
```

## From Github
1. Make sure you have an up-to-date version of cython, numpy, scipy and MDAnalysis:

``` 
pip install --user --upgrade cython numpy scipy MDAnalysis
```

2. Download and install pytim

```
git clone https://github.com/Marcello-Sega/pytim.git
cd pytim
python setup.py install --user
```

## Setting the `PYTHONPATH` variable

If you install with the option `--user` (which you have to do if you don't have administrator rights) you shouldn't forget to tell python where to look for the module by setting the `PYTHONPATH` environment variable. 

Under Linux, you could do, for example:
```
export PYTHONPATH=$HOME/.local/lib/python2.7/site-packages
```

Under OS-X, instead, use something like:
```
export PYTHONPATH=$HOME/Library/Python/2.7/lib/python/site-packages
```

You can search for the exact path by issuing the following command:

```
python -c "import site; print(site.USER_SITE)"
```

If this for some reason doesn't work, get some hint using:

```
find $HOME -name site-packages
```


## Trouble installing? 

Some of the most common issues are the following:

**Problem**: The system does not let me write (even using `sudo`) some files

**Solution**: You're most likely running under a recent version of OS-X. Always install packages as user (`pip install <package> --user`

**Problem**: cKDTree complains about the `boxsize` parameter

**Solution**: the version of `scipy` must be >= 0.18


**Problem**: Even though I've upgraded `scipy`, I keep getting problems about `boxsize`

**Solution**: You should tell python where to find packages by setting the variable `$PYTHONPATH` 

**Problem**: some error message mentioning `... file was built for x86_64 which is not the architecture being linked (i386)`

**Solution**: use `export ARCHFLAGS='-arch x86_64'` before installing

**Problem**: I'm getting an annoying message like "UserWarning: Module pytim_dbscan was already imported from [...]"

**Solution**: You've installed pytim, and are launching python within the pytim package directory. Move away from there :)



---------------------------

# References  <img src="https://raw.githubusercontent.com/Marcello-Sega/gitim/ITIM/media/soot1small.png" width="180" align="right" style="z-index:999;"> <a name="refs">

The [pytim paper is avaliable](https://onlinelibrary.wiley.com/doi/epdf/10.1002/jcc.25384)  (under the terms of the Creative Commons BY-NC 4.0 licence) from Wiley. Please cite it if you use pytim for your research:

[M. Sega, S. G. Hantal, B. Fabian and P. Jedlovszky, _J. Comp. Chem._ **39**, 2118-2125 (2018)](http://dx.doi.org/10.1002/jcc.25384) Pytim: A python package for the interfacial analysis of molecular simulations


```
@article{sega2018pytim,
  title={Pytim: A python package for the interfacial analysis of molecular simulations},
  author={Sega, M. and Hantal, G. and F{\'a}bi{\'a}n, B. and Jedlovszky, P.},
  journal={J. Comput. Chem.},
  pages={2118--2125},
  volume={39},
  year={2018},
  publisher={Wiley Online Library}
}
```

Depending on which algorithm you are using, you might also want to cite the following: 

[M. Sega, S. S. Kantorovich, P. Jedlovszky and M. Jorge, _J. Chem. Phys._ **138**, 044110 (2013)](http://dx.doi.org/10.1063/1.4776196) The generalized identification of truly interfacial molecules (ITIM) algorithm for nonplanar interfaces.

[L. B. Pártay, G. Hantal, P. Jedlovszky, Á. Vincze and G. Horvai, _J. Comp. Chem._ **29**, 945 (2008)](http://dx.doi.org/10.1002/jcc.20852) A new method for determining the interfacial molecules and characterizing the surface roughness in computer simulations. Application to the liquid–vapor interface of water

[M. Sega and G. Hantal._Phys. Chem. Chem. Phys._ **29**, 18968-18974 (2017)](https://doi.org/10.1039/C7CP02918G) Phase and interface determination in computer simulations of liquid mixtures with high partial miscibility.

[E. Chacón, P. Tarazona, Phys. Rev. Lett. **91**, 166103 (2003)](http://dx.doi.org/10.1103/PhysRevLett.91.166103) Intrinsic profiles beyond the capillary wave theory: A Monte Carlo study.

[A. P. Willard, D. Chandler, J. Phys. Chem. B **114**,1954 (2010)](http://dx.doi.org/10.1021/jp909219k) Instantaneous Liquid Interfaces



CHANGELOG.md
# Changelog
Changelog for the pytim project

## [Unreleased]
### Changed

- Added DistributionFunction observable in distributionfunction.py. 
  This is aims at being a class from which one can derive both density 
  profiles/maps and correlation functions.

- Added ReferenceFrame observable in basic_observables.py



pytim/itim.py
#!/usr/bin/python
# -*- Mode: python; tab-width: 4; indent-tabs-mode:nil; coding: utf-8 -*-
# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4
""" Module: itim
    ============
"""

from __future__ import print_function
from multiprocessing import Process, Queue
import numpy as np
try:
    from __builtin__ import zip as builtin_zip
except:
    from builtins import zip as builtin_zip
from scipy.spatial import cKDTree

from . import messages
from . import utilities
from .surface import SurfaceFlatInterface as Surface
from .sanity_check import SanityCheck

from .interface import Interface
from .patches import patchTrajectory, patchOpenMM, patchMDTRAJ


class ITIM(Interface):
    """ Identifies interfacial molecules at macroscopically flat interfaces.

        *(Pártay, L. B.; Hantal, Gy.; Jedlovszky, P.; Vincze, Á.; Horvai, G., \
J. Comp. Chem. 29, 945, 2008)*

        :param Object universe:   The MDAnalysis_ Universe, MDTraj_ trajectory
                                  or OpenMM_ Simulation objects.
        :param Object group:      An AtomGroup, or an array-like object with
                                  the indices of the atoms in the group.  Will
                                  identify the interfacial molecules from this
                                  group
        :param float alpha:       The probe sphere radius
        :param str normal:        The macroscopic interface normal direction
                                  'x','y', 'z' or 'guess' (default)
        :param bool molecular:    Switches between search of interfacial
                                  molecules / atoms (default: True)
        :param int max_layers:    The number of layers to be identified
        :param dict radii_dict:   Dictionary with the atomic radii of the
                                  elements in the group. If None is supplied,
                                  the default one (from GROMOS 43a1) will be
                                  used.
        :param float cluster_cut: Cutoff used for neighbors or density-based
                                  cluster search (default: None disables the
                                  cluster analysis)
        :param float cluster_threshold_density: Number density threshold for
                                  the density-based cluster search. 'auto'
                                  determines the threshold automatically.
                                  Default: None uses simple neighbors cluster
                                  search, if cluster_cut is not None
        :param Object extra_cluster_groups: Additional groups, to allow for
                                  mixed interfaces
        :param bool info:         Print additional info
        :param bool centered:     Center the  :py:obj:`group`
        :param bool warnings:     Print warnings
        :param float mesh:        The grid spacing used for the testlines
                                  (default 0.4 Angstrom)
        :param bool autoassign:   If true (default) detect the interface
                                  every time a new frame is selected.

        Example:

        >>> import MDAnalysis as mda
        >>> import numpy as np
        >>> import pytim
        >>> from pytim.datafiles import *
        >>>
        >>> u = mda.Universe(WATER_GRO)
        >>> oxygens = u.select_atoms("name OW")
        >>>
        >>> interface = pytim.ITIM(u, alpha=1.5, max_layers=4,molecular=True)

        >>> # atoms in the layers can be accesses either through
        >>> # the layers array:
        >>> print (interface.layers)
        [[<AtomGroup with 786 atoms> <AtomGroup with 681 atoms>
          <AtomGroup with 663 atoms> <AtomGroup with 651 atoms>]
         [<AtomGroup with 786 atoms> <AtomGroup with 702 atoms>
          <AtomGroup with 666 atoms> <AtomGroup with 636 atoms>]]


        >>> interface.layers[0,0] # upper side, first layer
        <AtomGroup with 786 atoms>

        >>> interface.layers[1,2] # lower side, third layer
        <AtomGroup with 666 atoms>

        >>> # or as a whole AtomGroup. This can include all atoms in all layers
        >>> interface.atoms
        <AtomGroup with 5571 atoms>


        >>> selection = interface.atoms.sides == 0
        >>> interface.atoms[ selection ] # all atoms in the upper side layer
        <AtomGroup with 2781 atoms>
        >>> selection = np.logical_and(interface.atoms.layers == 2 , selection)
        >>> interface.atoms[ selection ] # upper side, second layer
        <AtomGroup with 681 atoms>

        >>> # the whole system can be quickly saved to a pdb file
        >>> # including the layer information, written in the beta field
        >>> # using:
        >>> interface.writepdb('system.pdb',centered=True)

        >>> # of course, the native interface of MDAnalysis can be used to
        >>> # write pdb files, but the centering options are not available.
        >>> # Writing to other formats that do not support the beta factor
        >>> # will loose the information on the layers.
        >>> interface.atoms.write('only_layers.pdb')

        >>> # In some cases it might be necessary to compute two interfaces.
        >>> # This could be done in the following way:
        >>> import MDAnalysis as mda
        >>> import pytim
        >>> from pytim.datafiles import WATER_GRO, WATER_XTC
        >>> u = mda.Universe(WATER_GRO,WATER_XTC)
        >>> u2 = mda.Universe(WATER_GRO,WATER_XTC)
        >>> inter = pytim.ITIM(u,group=u.select_atoms('resname SOL'))
        >>> inter2 = pytim.ITIM(u2,group=u2.select_atoms('resname SOL'))
        >>> for ts in u.trajectory[::50]:
        ... 	ts2 = u2.trajectory[ts.frame]

        .. _MDAnalysis: http://www.mdanalysis.org/
        .. _MDTraj: http://www.mdtraj.org/
        .. _OpenMM: http://www.openmm.org/
    """

    @property
    def layers(self):
        """Access the layers as numpy arrays of AtomGroups.

        The object can be sliced as usual with numpy arrays, so, for example:

        >>> import MDAnalysis as mda
        >>> import pytim
        >>> from pytim.datafiles import *
        >>>
        >>> u = mda.Universe(WATER_GRO)
        >>> oxygens = u.select_atoms("name OW")
        >>>
        >>> interface = pytim.ITIM(u, alpha=1.5, max_layers=4,molecular=True)
        >>> print(interface.layers[0,:])  # upper side (0), all layers
        [<AtomGroup with 786 atoms> <AtomGroup with 681 atoms>
         <AtomGroup with 663 atoms> <AtomGroup with 651 atoms>]


        >>> repr(interface.layers[1,0])  # lower side (1), first layer (0)
        '<AtomGroup with 786 atoms>'


        >>> print(interface.layers[:,0:3]) # 1st - 3rd layer (0:3), on both sides
        [[<AtomGroup with 786 atoms> <AtomGroup with 681 atoms>
          <AtomGroup with 663 atoms>]
         [<AtomGroup with 786 atoms> <AtomGroup with 702 atoms>
          <AtomGroup with 666 atoms>]]


        >>> print(interface.layers[1,0:4:2]) # side 1, layers 1-4 & stride 2 (0:4:2)
        [<AtomGroup with 786 atoms> <AtomGroup with 666 atoms>]

        """

        return self._layers

    def __init__(self,
                 universe,
                 group=None,
                 alpha=1.5,
                 normal='guess',
                 molecular=True,
                 max_layers=1,
                 radii_dict=None,
                 cluster_cut=None,
                 cluster_threshold_density=None,
                 extra_cluster_groups=None,
                 info=False,
                 centered=False,
                 warnings=False,
                 mesh=0.4,
                 autoassign=True,
                 **kargs):

        self.autoassign = autoassign

        self.symmetry = 'planar'
        self.do_center = centered

        sanity = SanityCheck(self, warnings=warnings)
        sanity.assign_universe(universe, group)
        sanity.assign_alpha(alpha)
        sanity.assign_mesh(mesh)

        self.max_layers = max_layers
        self._layers = np.empty(
            [2, max_layers], dtype=self.universe.atoms[0].__class__)
        self._surfaces = np.empty(max_layers, dtype=type(Surface))
        self.info = info
        self.normal = None
        self.PDB = {}
        self.molecular = molecular

        sanity.assign_cluster_params(cluster_cut,
                                     cluster_threshold_density, extra_cluster_groups)
        sanity.assign_normal(normal)
        sanity.assign_radii(radii_dict=radii_dict)

        self.grid = None
        self.use_threads = False

        patchTrajectory(self.universe.trajectory, self)

        self._assign_layers()

    def _create_mesh(self):
        """ Mesh assignment method

            Based on a target value, determine a mesh size for the testlines
            that is compatible with the simulation box.
            Create the grid and initialize a cKDTree object with it to
            facilitate fast searching of the gridpoints touched by molecules.
        """
        box = utilities.get_box(self.universe, self.normal)
        n, d = utilities.compute_compatible_mesh_params(self.target_mesh, box)
        self.mesh_nx = n[0]
        self.mesh_ny = n[1]
        self.mesh_dx = d[0]
        self.mesh_dy = d[1]

        _x = np.linspace(0, box[0], num=self.mesh_nx, endpoint=False)
        _y = np.linspace(0, box[1], num=self.mesh_ny, endpoint=False)
        _X, _Y = np.meshgrid(_x, _y)
        self.meshpoints = np.array([_X.ravel(), _Y.ravel()]).T
        self.meshtree = cKDTree(self.meshpoints, boxsize=box[:2])

    def _touched_lines(self, atom, _x, _y, _z, _radius):
        return self.meshtree.query_ball_point([_x[atom], _y[atom]],
                                              _radius[atom] + self.alpha)

    def _append_layers(self, uplow, layer, layers):
        inlayer_indices = np.flatnonzero(self._seen[uplow] == layer + 1)
        inlayer_group = self.cluster_group[inlayer_indices]

        if self.molecular is True:
            # we first select the (unique) residues corresponding
            # to inlayer_group, and then we create  group of the
            # atoms belonging to them, with
            # inlayer_group.residues.atoms
            inlayer_group = inlayer_group.residues.atoms

            # now we need the indices within the cluster_group,
            # of the atoms in the molecular layer group;
            # NOTE that from MDAnalysis 0.16, .ids runs from 1->N
            # (was 0->N-1 in 0.15), we use now .indices
            indices = np.flatnonzero(
                np.in1d(self.cluster_group.atoms.indices,
                        inlayer_group.atoms.indices))
            # and update the tagged, sorted atoms
            self._seen[uplow][indices] = layer + 1

        # one of the two layers (upper,lower) or both are empty
        if not inlayer_group:
            raise Exception(messages.EMPTY_LAYER)

        layers.append(inlayer_group)

    def _assign_one_side(self,
                         uplow,
                         sorted_atoms,
                         _x,
                         _y,
                         _z,
                         _radius,
                         queue=None):
        layers = []
        for layer in range(0, self.max_layers):
            # this mask tells which lines have been touched.
            mask = self.mask[uplow][layer]
            # atom here goes to 0 to #sorted_atoms, it is not a MDAnalysis
            # index/atom
            for atom in sorted_atoms:
                if self._seen[uplow][atom] != 0:
                    continue

                touched_lines = self._touched_lines(atom, _x, _y, _z, _radius)
                _submask = mask[touched_lines]

                if (len(_submask[_submask == 0]) == 0):
                    # no new contact, let's move to the next atom
                    continue

                # let's mark now:  1) the touched lines
                mask[touched_lines] = 1

                # 2) the sorted atoms.
                self._seen[uplow][atom] = layer + 1

                # 3) if all lines have been touched, create a group that
                # includes all atoms in this layer
                if np.sum(mask) == len(mask):
                    self._append_layers(uplow, layer, layers)
                    break
        if (queue is None):
            return layers
        else:
            queue.put(layers)

    def _prepare_layers_assignment(self):
        self._create_mesh()
        size = (2, int(self.max_layers), int(self.mesh_nx) * int(self.mesh_ny))
        self.mask = np.zeros(size, dtype=int)
        self.prepare_box()

    def _prelabel_groups(self):
        # first we label all atoms in group to be in the gas phase
        self.label_group(self.analysis_group.atoms, beta=0.5)
        # then all atoms in the largest group are labelled as liquid-like
        self.label_group(self.cluster_group.atoms, beta=0.0)

    def _assign_layers(self):
        """ Determine the ITIM layers.

            Note that the multiproc option is mainly for debugging purposes:
            >>> import MDAnalysis as mda
            >>> import pytim
            >>> u = mda.Universe(pytim.datafiles.WATER_GRO)
            >>> inter = pytim.ITIM(u,multiproc=True)
            >>> test1 = len(inter.layers[0,0])
            >>> inter = pytim.ITIM(u,multiproc=False)
            >>> test2 = len(inter.layers[0,0])
            >>> test1==test2
            True

        """
        up, low = 0, 1

        self.reset_labels()

        self._prepare_layers_assignment()
        # groups have been checked already in _sanity_checks()

        self._define_cluster_group()

        # we always (internally) center in ITIM
        self.center(planar_to_origin=True)

        self._prelabel_groups()

        _radius = self.cluster_group.radii
        size = len(self.cluster_group.positions)
        self._seen = [
            np.zeros(size, dtype=np.int8),
            np.zeros(size, dtype=np.int8)
        ]

        _x = utilities.get_x(self.cluster_group, self.normal)
        _y = utilities.get_y(self.cluster_group, self.normal)
        _z = utilities.get_z(self.cluster_group, self.normal)

        sort = np.argsort(_z + _radius * np.sign(_z))
        # NOTE: np.argsort returns the sorted *indices*

        # so far, it justs exploit a simple scheme splitting
        # the calculation between the two sides. Would it be
        # possible to implement easily 2d domain decomposition?
        proc, queue = [None, None], [Queue(), Queue()]
        proc[up] = Process(
            target=self._assign_one_side,
            args=(up, sort[::-1], _x, _y, _z, _radius, queue[up]))
        proc[low] = Process(
            target=self._assign_one_side,
            args=(low, sort[::], _x, _y, _z, _radius, queue[low]))

        for p in proc:
            p.start()
        for uplow in [up, low]:
            for index, group in enumerate(queue[uplow].get()):
                # cannot use self._layers[uplow][index] = group, otherwise
                # info about universe is lost (do not know why yet)
                # must use self._layers[uplow][index] =
                # self.universe.atoms[group.indices]
                self._layers[uplow][index] = self.universe.atoms[group.indices]
        for p in proc:
            p.join()
        for q in queue:
            q.close()

        self.label_planar_sides()

        for nlayer, layer in enumerate(self._layers[
                0]):  # TODO should this be moved out of assign_layers?
            self._surfaces[nlayer] = Surface(self, options={'layer': nlayer})

        if self.do_center is False:  # NOTE: do_center requires centering in
            # the middle of the box.
            # ITIM always centers internally in the
            # origin along the normal.
            self.universe.atoms.positions = self.original_positions
        else:
            self._shift_positions_to_middle()


#
